### DISCUSSION TITLE
Questions about the QuestionnaireStepper widget

### DISCUSSION THREAD

**Discussion by bennyhaeusler** [sujrd/faiadashu#113] [Q&A] [answered]
Hello there,

first of all I have to mention that you did a great job in updating, fixing and improving the original faiadashu package. This is amazing!

We used the old original package in one of our research projects, unfortunately the maintainer seems to lack the time to continue it's development. I read the [issue](https://github.com/tiloc/faiadashu/issues/17) which lead me to your fork.

I've been testing your overhauled implementation of the QuestionnaireStepper widget with some of our questionnaires and ran into a couple of questions:

1. If an item is of type "group" but does not have a "text" set, an empty page is rendered. Is this behaviour intentional or am I missing some settings? I can manually skip these "empty pages" when I hook into the onBeforeChanged callback provided by QuestionnaireStepperPageViewData. Though the callback seems to not be invoked when navigating back to a previous page ( calling the previousPage() method on a QuestionnaireStepperPageViewController). 

Here is an example where the first page is empty:
```json
{
  "resourceType": "Questionnaire",
  "title": "Testing QuestionnaireStepper",
  "status": "draft",
  "item": [
    {
      "linkId": "firstGroup",
      "type": "group",
      "item": [
        {
          "linkId": "textItem",
          "type": "text",
          "text": "enter some text"
        }
      ]
    }
  ]
}
```

2. When there is an item of type "group" with one (or more) subitems of type "display" (and no other subitems of any other type) and I call nextPage() on the QuestionnaireStepperPageViewController, the QuestionnaireStepper widget does not navigate to the next page, because _state?.isUserInteractionAllowed() evaluates to false within the nextPage() method of the QuestionnaireStepperPageViewController class.

Here is an example:
```json
{
  "resourceType": "Questionnaire",
  "title": "Testing QuestionnaireStepper",
  "status": "draft",
  "item": [
    {
      "linkId": "firstGroup",
      "type": "group",
      "item": [
        {
          "linkId": "displayItem1",
          "type": "display",
          "text": "hello there"
        },
        {
          "linkId": "displayItem2",
          "type": "display",
          "text": "hello there again"
        }
      ]
    },
    {
      "linkId": "textItem",
      "type": "text",
      "text": "enter some text"
    }
  ]
}
```

Can you help me on how to handle these two cases?

Kind regards,
Benny

**Comment by hannyprastya** [comment, marked as answer]
Thank you for taking an interest in this project.

> If an item is of type "group" but does not have a "text" set, an empty page is rendered. Is this behaviour intentional or am I missing some settings?

Yes, this is intentional, but we could definitely improve it. We welcome a proposal or even an PR :wink:

Another possible solution, you can override the behaviors by creating an instance of `QuestionnaireThemeData` [here](https://github.com/sujrd/faiadashu/blob/ea430b7c13e8c3b099c1b24bae4d999560ec31d4/lib/questionnaires/view/src/questionnaire_theme.dart#L36). 

> Though the callback seems to not be invoked when navigating back to a previous page ( calling the previousPage() method on a QuestionnaireStepperPageViewController).

Thank you for noticing this, this is not intended, I will create an issue to address this.

> When there is an item of type "group" with one (or more) subitems of type "display" (and no other subitems of any other type) and I call nextPage() on the QuestionnaireStepperPageViewController, the QuestionnaireStepper widget does not navigate to the next page, because _state?.isUserInteractionAllowed() evaluates to false within the nextPage() method of the QuestionnaireStepperPageViewController class.

Again, thank you for the report. I will try to find time to investigate this.

**Reply by bennyhaeusler** [reply]
Great, thank you for the feedback! 
I will follow your suggestion and check out the QuestionnaireThemeData :-)

Kind regards,
Benny
